src/common/utils/ldap/ldap.go

```diff
@@ -0,0 +1,367 @@
+/*
+   Copyright (c) 2016 VMware, Inc. All Rights Reserved.
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package ldap
+
+import (
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"crypto/tls"
+
+	"github.com/vmware/harbor/src/common/dao"
+	"github.com/vmware/harbor/src/common/models"
+	"github.com/vmware/harbor/src/common/utils/log"
+	"github.com/vmware/harbor/src/ui/config"
+
+	goldap "gopkg.in/ldap.v2"
+)
+
+var attributes = []string{"uid", "cn", "mail", "email"}
+
+// GetSystemLdapConf ...
+func GetSystemLdapConf() (models.LdapConf, error) {
+	var err error
+	var ldapConfs models.LdapConf
+
+	if config.AuthMode() != "ldap_auth" {
+		return ldapConfs, fmt.Errorf("system auth_mode isn't ldap_auth, please check configuration")
+	}
+
+	ldapConfs.LdapURL = config.LDAP().URL
+	ldapConfs.LdapSearchDn = config.LDAP().SearchDn
+	ldapConfs.LdapSearchPassword = config.LDAP().SearchPwd
+	ldapConfs.LdapBaseDn = config.LDAP().BaseDn
+	ldapConfs.LdapFilter = config.LDAP().Filter
+	ldapConfs.LdapUID = config.LDAP().UID
+	ldapConfs.LdapScope, err = strconv.Atoi(config.LDAP().Scope)
+	if err != nil {
+		log.Errorf("invalid LdapScope format from system, error: %v", err)
+		return ldapConfs, err
+	}
+
+	ldapConfs.LdapConnectionTimeout, err = strconv.Atoi(config.LDAP().ConnectTimeout)
+	if err != nil {
+		log.Errorf("invalid LdapConnectionTimeout format from system, error: %v", err)
+		return ldapConfs, err
+	}
+
+	return ldapConfs, nil
+
+}
+
+// ValidateLdapConf ...
+func ValidateLdapConf(ldapConfs models.LdapConf) (models.LdapConf, error) {
+	var err error
+
+	if ldapConfs.LdapURL == "" {
+		return ldapConfs, fmt.Errorf("can not get any available LDAP_URL")
+	}
+
+	ldapConfs.LdapURL, err = formatLdapURL(ldapConfs.LdapURL)
```

Comment:
empty string  is an malformed ldapURL and it's an error should be returned by "formatLdapURL"